// pry/lib/pry/commands/help.js
(function() {
  var Command, Help, Range, chalk,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Command = require('../command');

  Range = require('../range');

  chalk = require('chalk');

  Help = (function(superClass) {
    extend(Help, superClass);

    function Help() {
      return Help.__super__.constructor.apply(this, arguments);
    }

    Help.prototype.name = 'help';

    Help.prototype.aliases = ['\\?'];

    Help.prototype.definition = 'Shows a list of commands. Type `help foo` for help on the `foo` command.';

    Help.prototype.args = new Range(0, 1);

    Help.prototype.typeahead = function(input) {
      var command, items, name, ref;
      if (input == null) {
        input = '';
      }
      if (input.indexOf('help') === 0) {
        items = [];
        ref = this.commands();
        for (name in ref) {
          command = ref[name];
          if (command.name) {
            items.push("help " + command.name);
          }
        }
        return items;
      } else {
        return ['help'];
      }
    };

    Help.prototype.execute = function(arg, chain) {
      var command, e, name, ref;
      name = arg[0];
      try {
        if (name) {
          command = this.command(name);
          this.output.add(chalk.blue(command.name), '-', command.definition);
          this.output.sendAll();
        } else {
          this.output.add(chalk.blue('mode'), '-', 'toggle between coffee and javascript modes');
          ref = this.commands();
          for (name in ref) {
            command = ref[name];
            if (command.name) {
              this.output.add(chalk.blue(command.name), '-', command.definition);
            }
          }
          this.output.sendAll();
        }
      } catch (_error) {
        e = _error;
        console.log("Can't do it.");
      }
      return chain.next();
    };

    return Help;

  })(Command);

  module.exports = Help;

}).call(this);
