// pry/lib/pry/command.js
(function() {
  var Command, Range;

  Range = require('./range');

  Command = (function() {
    Command.commands = {};

    Command.prototype.name = '';

    Command.prototype.aliases = [];

    Command.prototype.definition = '';

    Command.prototype.help = '';

    Command.prototype.args = new Range(0, 0);

    function Command(arg) {
      this.scope = arg.scope, this.output = arg.output, this.prompt = arg.prompt, this.app = arg.app;
      this.constructor.commands[this.constructor.name] = this;
    }

    Command.prototype.command = function(input) {
      var command, name, ref;
      ref = this.commands();
      for (name in ref) {
        command = ref[name];
        if (command.constructor.name.match(new RegExp(input, 'i'))) {
          return command;
        }
      }
    };

    Command.prototype.commands = function() {
      return this.constructor.commands;
    };

    Command.prototype.typeahead = function() {
      var items;
      items = this.aliases.slice(0);
      items.push(this.name);
      return items;
    };

    Command.prototype.command_regex = function() {
      var subject;
      subject = "^(?:" + this.name;
      if (this.aliases.length > 0) {
        subject += "|" + (this.aliases.join('|'));
      }
      subject += ")((?: (?:[^ ]+))" + (this.args.to_regex()) + ")$";
      return new RegExp(subject);
    };

    Command.prototype.match = function(input) {
      return input.match(this.command_regex());
    };

    Command.prototype.find_file = function() {
      return this.app.find_file();
    };

    return Command;

  })();

  module.exports = Command;

}).call(this);
