// pry/lib/pry/compiler.js
(function() {
  var Compiler, coffee, pry;

  coffee = require('coffee-script');

  pry = require('../pry');

  Compiler = (function() {
    Compiler.prototype.mode_id = 0;

    Compiler.prototype.modes = ['js', 'coffee'];

    function Compiler(arg) {
      var isCoffee;
      this.scope = arg.scope, isCoffee = arg.isCoffee;
      if (isCoffee) {
        this.mode_id = 1;
      }
    }

    Compiler.prototype.mode = function() {
      return this.modes[this.mode_id];
    };

    Compiler.prototype.toggle_mode = function() {
      return this.mode_id = (this.mode_id + 1) % this.modes.length;
    };

    Compiler.prototype.execute = function(code, language) {
      if (language == null) {
        language = this.modes[this.mode_id];
      }
      return this["execute_" + language](code);
    };

    Compiler.prototype.execute_coffee = function(code) {
      var linesOfJs;
      linesOfJs = coffee.compile(code, {
        bare: true
      }).split("\n");
      return this.execute_js(linesOfJs.filter(function(l) {
        return l.length > 0 && l.trim().slice(0, 3) !== 'var';
      }).join("\n"));
    };

    Compiler.prototype.execute_js = function(code) {
      return this.scope(code);
    };

    return Compiler;

  })();

  module.exports = Compiler;

}).call(this);
