// pry/lib/pry/sync_prompt.js
(function() {
  var EventEmitter, MultilineState, SinglelineState, SyncPrompt, chalk, deasync, fs, readline,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  readline = require('readline');

  fs = require('fs');

  EventEmitter = (require('events')).EventEmitter;

  deasync = require('deasync');

  chalk = require('chalk');

  MultilineState = (function() {
    function MultilineState() {}

    MultilineState.prototype.data = '';

    MultilineState.prototype.keypress = function(input, chars) {
      this.data += chars;
      if (this.data.match(/(\r|\n)\1$/)) {
        this.data = '';
        input.state('single');
        return input.send_data();
      } else if (chars.match(/(\r|\n)$/)) {
        return input.prompt();
      }
    };

    MultilineState.prototype.prompt = function(input, prompt) {
      if (this.data === '') {
        input.cli.setPrompt(prompt.replace(/[^>](?!$)/g, '-'));
      } else {
        input.cli.setPrompt(prompt.replace(/.(?!$)/g, '.'));
      }
      return input.cli._refreshLine();
    };

    return MultilineState;

  })();

  SinglelineState = (function() {
    function SinglelineState() {}

    SinglelineState.prototype.keypress = function(input, chars) {
      if (chars === '\u0016') {
        input.state('multi');
        return input.prompt();
      } else if (chars.match(/(\r|\n)$/)) {
        return input.send_data();
      }
    };

    SinglelineState.prototype.prompt = function(input, prompt) {
      input.cli.setPrompt(prompt);
      return input.cli.prompt();
    };

    return SinglelineState;

  })();

  SyncPrompt = (function(superClass) {
    extend(SyncPrompt, superClass);

    SyncPrompt.prototype.lines = '';

    SyncPrompt.prototype.count = 0;

    SyncPrompt.prototype.states = {
      multi: new MultilineState,
      single: new SinglelineState
    };

    SyncPrompt.prototype._state = 'single';

    SyncPrompt.prototype.done = false;

    function SyncPrompt(arg) {
      var hist, lastLine, typeahead;
      typeahead = arg.typeahead, this.mode = arg.mode;
      this.close = bind(this.close, this);
      this.type = bind(this.type, this);
      this.prompt = bind(this.prompt, this);
      this.send_data = bind(this.send_data, this);
      this.keypress = bind(this.keypress, this);
      this.line = bind(this.line, this);
      this.state = bind(this.state, this);
      this.indent = '';
      this.cli = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        completer: typeahead
      });
      try {
        hist = fs.readFileSync(process.env.HOME + "/.pryjs_history").toString();
        this.cli.history = hist.split('\n').reverse();
      } catch (_error) {
        null;
      }
      lastLine = this.cli.history[0];
      this.cli.on('line', (function(_this) {
        return function(line) {
          if (line && line.length && line !== lastLine) {
            fs.appendFile(process.env.HOME + "/.pryjs_history", "\n" + line);
            lastLine = line;
          }
          return _this.line(line);
        };
      })(this));
      process.stdin.on('data', this.keypress);
    }

    SyncPrompt.prototype.state = function(state) {
      if (state) {
        this._state = state;
      }
      return this.states[this._state];
    };

    SyncPrompt.prototype.line = function(line) {
      if (line.charCodeAt(0) === 22) {
        line = line.slice(1);
      }
      return this.lines += '\n' + line;
    };

    SyncPrompt.prototype.keypress = function(chars) {
      return this.state().keypress(this, chars.toString());
    };

    SyncPrompt.prototype.send_data = function() {
      this.count++;
      this.emit('data', this.lines.trim(), {
        next: this.prompt,
        stop: this.close
      });
      return this.lines = '';
    };

    SyncPrompt.prototype.prompt = function() {
      return this.state().prompt(this, ["[" + this.count + "] ", this.mode === 'js' ? chalk.white('pryjs') : chalk.blue('pryjs'), this.indent ? "* " + this.indent : '> '].join(''));
    };

    SyncPrompt.prototype.open = function() {
      var results;
      this.done = false;
      this.prompt();
      results = [];
      while (!this.done) {
        results.push(deasync.runLoopOnce());
      }
      return results;
    };

    SyncPrompt.prototype.type = function(input) {
      this.lines = input;
      return this.send_data();
    };

    SyncPrompt.prototype.close = function() {
      this.done = true;
      process.stdin.removeListener('data', this.keypress);
      return this.cli.close();
    };

    return SyncPrompt;

  })(EventEmitter);

  module.exports = SyncPrompt;

}).call(this);
